Tweaks to cmake build

wavexx opened the pull request:
Giving this a go and tweaking a few things along the way.

Being able to build more than a single variant in the same tree is.. intriguing. Didn't think of doing it this way.

I would have some further proposals:

- move utils/bootstrap.py to tools/, since we already have it. Or do the opposite (so that we have one generic "tooldir").
- strip -E3Dv6full and 1_75_mm from the variant file names directly (that is, by renaming them). 

It's ok to strip them like we do, but I want to avoid dumb filename conflicts if you just *happen* to have a second pre-stripped filename in there.

@3d-gussner would renaming 1_75mm_MK3S-EINSy10a-E3Dv6full.h to MK3S-EINSy10a be an issue?

wavexx committed:
lcd_print_pad: do not overflow len when truncating the string

wavexx committed:
Merge pull request #3628 from wavexx/lcd_pad_fix

lcd_print_pad: do not overflow len when truncating the string

3d-gussner committed:
BETA firmware message only in English
Rephrase BETA firmware message to fit on two screens
ALPHA message same as DEVEL
Fix ALPHA DEVEL message

3d-gussner committed:
Fix MSG_SPOOL_JOIN

3d-gussner committed:
Rename EN_ONLY to EN_FARM
remove DEBUG/DEVEL/ALPHA/BETA lcd warning

3d-gussner committed:
Merge pull request #3633 from 3d-gussner/MK3_BETA_message

Remove BETA Warnings

wavexx committed:
cleanup: Supress missing return warnings

Rewrite the function with a temporary to avoid the spourious no-return
warning.

The generated code is *unchanged*.

wavexx committed:
cleanup: Do not return const values from functions

This generates a warning under -Wextra, since in most cases a const
value doesn't prevent buggy code (as copies are allowed) while
preventing some optimizations (such as move operations) to take place.

wavexx committed:
cleanup: Reorder members to match initialization order

wavexx committed:
cleanup: Remove return value from MMU2Serial::write()

Removes a missing return warning.

Looks like we don't handle overflow at the moment and in all surrounding
code, so let's remove the return value for now.

wavexx committed:
cleanup: Remove implicit fallthrough in CRC validation

wavexx committed:
cleanup: Suppress unused argument warnings

wavexx committed:
Merge pull request #3639 from wavexx/cleanup_warnings

Cleanup warnings

wavexx committed:
cmake: Do not set PROJECT_VERSION in PARENT_SCOPE

wavexx committed:
cmake: Remove obsolete tests

wavexx committed:
cmake: Update Catch2

wavexx committed:
cmake: Move tests and CMakefiles to tests/

wavexx committed:
cmake: Allow to configure/restrict the variant to be built

This introduces FW_VARIANTS as a configurable option, defaulting to the
full list of variants.

3d-gussner commented:
@wavexx I think the `1_75mm_...-E3Dv6full` is for the past as there have been [2.85mm and v5 hotends](https://help.prusa3d.com/guide/introduction_132507).
I think we can rename them as they are only used in the source code and the build scripts. 

- [ ] Update `build.sh`
- [ ] Update `PF-build.sh`
- [ ] Update `.travis.yml`
- [ ] Update `build server scripts`
- [ ] Update `README.md`

3d-gussner commented on tests/CMakeLists.txt:
```diff
@@ -0,0 +1,19 @@
+include(${Catch2_SOURCE_DIR}/contrib/Catch.cmake)
+
+# Make test executable
+set(TEST_SOURCES
```
These have been removed from `.travis.yml`

wavexx commented:
Does the build server rely on the name somehow?

3d-gussner commented:
The build server uses the different variants files and these are "hard-coded". But easy to change.

wavexx committed:
cmake: Tighten ignores

wavexx committed:
cmake: Uniform target compile/link flags

wavexx committed:
cmake: Make builds fully reproducible by default

- Work-around SOURCE flags being reset for each project when setting the
  random-seed property by calling a support function.
- Likewise, set correct flags for reproducibility when creating archives.
- Strip source prefix paths from debug info

This should set the EPOCH using the last git commit date when available,
but it's not currently implemented. Just set the EPOCH to be 0 for now.

wavexx committed:
cmake: Only make the target reproducible

wavexx committed:
cmake: Work-around gcc 7 not supporting SOURCE_DATE_EPOCH

wavexx commented:
Didn't check the sizes/languages yet, but it looks like that now, _finally_, the FW itself is reproducible, which I consider a major step forward.

Doesn't set a fancy date yet, but we'll get that from the last git commit somehow.

wavexx committed:
cmake: Generate correct inline source listings in asm output

Manually provide a source path prefix to objdump

wavexx committed:
cmake: Correctly split/gc sections while linking

Repeat the flags while linking for LTO

wavexx commented:
I'm also getting perfectly matching binary sizes now on the EN-only build.

wavexx commented:
Going to stop here for this PR ;)

vintagepc commented on CMakeLists.txt:
```diff
@@ -60,6 +60,7 @@ SET(SELECTED_LANGUAGES ${MAIN_LANGUAGES} ${COMMUNITY_LANGUAGES})
 
 get_dependency_directory(prusa3dboards PRUSA_BOARDS_DIR)
 project(Prusa-Firmware)
+add_subdirectory(lib)
```
This should probably be excluded if the compiler is avr.

vintagepc committed:
Merge pull request #3641 from vintagepc/build-with-cmake

Build the firmware with cmake

vintagepc commented on CMakeLists.txt:
```diff
@@ -173,12 +184,6 @@ SET(FW_SOURCES
-    set_property(SOURCE ${_FILE} APPEND_STRING PROPERTY COMPILE_FLAGS "-frandom-seed=${_BASE}.o")
-endforeach()
-
-
```
Why did you remove this? Now your compiler will use random seeds all the time...

vintagepc commented on CMakeLists.txt:
```diff
@@ -458,42 +477,29 @@ endfunction()
-	foreach(THIS_VAR IN LISTS FW_VARIANTS)
-		string(REPLACE ".h" "" TRIMMED_NAME "${THIS_VAR}")
-		message("Variant added: ${TRIMMED_NAME}")
-        string(REPLACE "-E3Dv6full" "" DIR_NAME "${TRIMMED_NAME}")
```
this was done so as not to break the existing builds... I'd leave it until we actually make the change, the prefix and postfix are just garbage/noise on screen and add to directory lengths/paths to anannoying degree.

vintagepc replied:
nevermind - stupid diff view is terrible sometimes.

wavexx commented:
Closing
